cli/connhelper/commandconn/commandconn.go

```diff
@@ -200,16 +212,32 @@ func (c *commandConn) CloseWrite() error {
 
 func (c *commandConn) Write(p []byte) (int, error) {
 	n, err := c.stdin.Write(p)
-	if err == io.EOF || errors.Is(err, os.ErrClosed) {
+	// check after the call to Write, since
+	// it is blocking, and while waiting on it
+	// Close might get called
+	if c.closing.Load() {
+		// If we're currently closing the connection
+		// we don't want to call onEOF, but we do want
+		// to return an io.EOF
+		return 0, io.EOF
+	}
+
+	if err == io.EOF {
 		err = c.onEOF(err)
 	}
 	return n, err
 }
 
 // Close is the net.Conn func that gets called
-// by the transport when a request is cancelled
-// cancelled due to it's context timing out
+// by the transport when a dial is cancelled
+// due to it's context timing out. Any blocked
+// Read or Write calls will be unblocked and
+// return errors. It will block until the underlying
+// command has terminated.
 func (c *commandConn) Close() error {
+	c.closing.Store(true)
+	defer c.closing.Store(false)
+
 	var err error
 	if err = c.CloseRead(); err != nil {
```

Comment:
Not new, but this code looks a bit iffy;

- `c.CloseRead()` is assigned to `err`, and logs it if it went wrong
- `c.CloseWrite()` is then unconditionally overwriting `err` (and logs if it went wrong)
- the function as a whole returns the last result (which could be `nil` if `c.CloseWrite()` succeeded, but `c.CloseRead()` failed

Wondering what the expected result is here;

- should we ignore the first error if the second was successful? 
- do we need to return an error if _either_ failed?
- do we need a `multiErr`, an errgroup (or variant thereof?)

Reply:
Ahhhh yeah, there's a flow here where we don't actually manage to terminate the command and yet return nil – I'll write a test for that, and change up the code.
If the 1st call fails and doesn't close `stdOut`, the 2nd call can work fine, close `stdIn`, and not call `kill()`, and then it doesn't return an error but we effectively haven't killed the connection.
This is probably fine 99% of the time (or more), but we can make it better for sure.

Reply:
Yes; this was all just from looking at it; I haven't considered what the consequences are for the incorrect handling (and perhaps it's okay for a follow up). 😅